README.md

```diff
@@ -18,36 +18,203 @@ testing first appeard in Erlang's proprietary QuickCheck. The
 `quickcheck-state-machine` library can be seen as an attempt to provide similar
 functionality to Haskell's QuickCheck library.
 
-### Sample run (teaser)
+### Example
 
-Here's a sample output from when we look for race conditions in the mutable
-reference example:
+Lets implement and test programs using mutable references. Our mutable
+references can be created, read from, written to and incremented:
 
+```haskell
+data Action (v :: * -> *) :: * -> * where
+  New   ::                                     Action v (Opaque (IORef Int))
+  Read  :: Reference v (Opaque (IORef Int)) -> Action v Int
+  Write :: Reference v (Opaque (IORef Int)) -> Int -> Action v ()
+  Inc   :: Reference v (Opaque (IORef Int)) -> Action v ()
 ```
-> quickCheck (MutableReference.prop_parallel RaceCondition)
-*** Failed! (after 5 tests and 6 shrinks):
 
-Couldn't linearise:
+In order to be able to show counterexamples, we need a show instance for our
+actions. `IORef`s don't have a show instance, thats why we wrap them in
+`Opaque`; which gives a show instance to a type that doesn't have one.
+
+When we generate actions we won't be able to create arbitrary `IORef`s, that's
+why all uses of `IORefs` are wrapped in `Reference v`, where the parameter `v`
+will let us use symbolic references while generating (and concrete ones when
+executing).
```

Comment:
I think it would be better to swap the order of these two paragraphs. I want the `Opaque` type to be more of a side-remark, but having it first makes it look like it is very important.